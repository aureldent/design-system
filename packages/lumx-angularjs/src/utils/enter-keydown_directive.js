import { ENTER_KEY_CODE } from '@lumx/core/js/constants';

/////////////////////////////

function EnterKeydownDirective() {
    'ngInject';

    function link(scope, el, attrs) {
        el.on('keydown', (evt) => {
            if (evt.which === ENTER_KEY_CODE) {
                scope.$apply(() => {
                    scope.$eval(attrs.lxEnterKeydown, { $event: evt });
                });

                evt.preventDefault();
            }
        });
    }

    return {
        link,
    };
}

/////////////////////////////

angular.module('lumx.utils.enter-keydown').directive('lxEnterKeydown', EnterKeydownDirective);

/////////////////////////////

export { EnterKeydownDirective };
