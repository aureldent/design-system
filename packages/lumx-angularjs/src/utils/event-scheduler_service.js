function EventSchedulerService($document, LxUtilsService) {
    'ngInject';

    const service = this;

    /////////////////////////////
    //                         //
    //    Private attributes   //
    //                         //
    /////////////////////////////

    /**
     * The handlers.
     *
     * @type {Object}
     */
    const _handlers = {};

    /**
     * The schedule.
     *
     * @type {Object}
     */
    const _schedule = {};

    /////////////////////////////
    //                         //
    //    Private functions    //
    //                         //
    /////////////////////////////

    /**
     * Handle en event.
     *
     * @param {Event} evt The event.
     */
    function _handle(evt) {
        let scheduler;

        for (const [key] of Object.entries(_schedule)) {
            if (key.includes(evt.type)) {
                scheduler = _schedule[key];
            }
        }

        if (angular.isDefined(scheduler)) {
            for (let i = 0, len = scheduler.length; i < len; i++) {
                const handler = scheduler[i];

                if (angular.isDefined(handler) && angular.isDefined(handler.cb) && angular.isFunction(handler.cb)) {
                    handler.cb(evt);

                    if (evt.isPropagationStopped()) {
                        break;
                    }
                }
            }
        }
    }

    /////////////////////////////
    //                         //
    //     Public functions    //
    //                         //
    /////////////////////////////

    /**
     * Register en event.
     *
     * @param  {string}   eventName The event name.
     * @param  {Function} cb        The event callback.
     * @return {string}   The event id.
     */
    function register(eventName, cb) {
        const handler = {
            cb,
            eventName,
        };

        const id = LxUtilsService.generateUUID();
        _handlers[id] = handler;

        if (angular.isUndefined(_schedule[eventName])) {
            _schedule[eventName] = [];

            $document.on(eventName, _handle);
        }

        _schedule[eventName].unshift(_handlers[id]);

        return id;
    }

    /**
     * Unregister en event.
     *
     * @param  {string} id The event id.
     * @return {found}  Whether the event has been found or not.
     */
    function unregister(id) {
        let found = false;
        const handler = _handlers[id];

        if (angular.isDefined(handler) && angular.isDefined(_schedule[handler.eventName])) {
            const index = _schedule[handler.eventName].indexOf(handler);

            if (angular.isDefined(index) && index > -1) {
                _schedule[handler.eventName].splice(index, 1);

                delete _handlers[id];
                found = true;
            }

            if (_schedule[handler.eventName].length === 0) {
                delete _schedule[handler.eventName];

                $document.off(handler.eventName, _handle);
            }
        }

        return found;
    }

    /////////////////////////////

    service.register = register;
    service.unregister = unregister;
}

/////////////////////////////

angular.module('lumx.utils.event-scheduler').service('LxEventSchedulerService', EventSchedulerService);

/////////////////////////////

export { EventSchedulerService };
