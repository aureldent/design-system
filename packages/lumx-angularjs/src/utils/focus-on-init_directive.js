import { COMPONENT_PREFIX, MODULE_NAME } from '@lumx/angularjs/constants/common_constants';

/////////////////////////////

function FocusOnInitDirective($timeout) {
    'ngInject';

    function link(scope, el, attrs) {
        if (angular.isDefined(attrs.lumxFocusOnInit) && attrs.lumxFocusOnInit && !scope.$eval(attrs.lumxFocusOnInit)) {
            return;
        }

        $timeout(() => {
            el.focus();
        });
    }

    return {
        link,
    };
}

/////////////////////////////

angular.module(`${MODULE_NAME}.utils.focus-on-init`).directive(`${COMPONENT_PREFIX}FocusOnInit`, FocusOnInitDirective);

/////////////////////////////

export { FocusOnInitDirective };
