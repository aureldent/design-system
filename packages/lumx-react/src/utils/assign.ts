export const assign = (target: any, ...sources: any[]) => {
    sources.forEach((source) => Object.keys(source).forEach((key) => (target[key] = source[key])));
    return target;
};
