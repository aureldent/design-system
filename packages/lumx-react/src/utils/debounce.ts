export function debounce(func: (...args: any) => any, debounceTime: number) {
    let timeout: any;
    return function debouncedFunction(...args: any[]) {
        clearTimeout(timeout);
        timeout = setTimeout(() => func(...args), debounceTime);
    };
}
