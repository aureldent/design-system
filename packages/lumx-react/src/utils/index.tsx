export { getBasicClass, handleBasicClasses, detectSwipe, onEnterPressed, onEscapePressed } from '@lumx/core/utils';

export * from './calculatePopoverPlacement';

export * from './isInViewPort';

export * from './validation';

export * from './partitionMulti';

export { getRootClassName } from './getRootClassName';

export * from './type';
