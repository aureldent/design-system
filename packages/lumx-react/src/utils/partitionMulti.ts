import { Predicate } from './type';
import reduce from 'lodash/reduce';
import concat from 'lodash/concat';
import dropRight from 'lodash/dropRight';
import partition from 'lodash/partition';
import last from 'lodash/last';

/**
 * Similar to lodash `partition` function but working with multiple predicates.
 *
 * @example
 * const isString = (s) => typeof s === 'string'
 * const isNumber = (s) => typeof s === 'number'
 * const [strings, numbers, others] = partitionMulti(['a', 1, 'b', false], [isString, isNumber])
 * //=> [['a', 'b'], [1], [false]]
 *
 * @param  elements array of elements
 * @param  predicates array of predicates to apply on elements
 * @return partitioned elements by the given predicates
 */
function partitionMulti<T>(elements: T[], predicates: Array<Predicate<T>>): T[][] {
    return reduce(
        predicates,
        (partitioned: T[][], predicate: Predicate<T>) =>
            concat(dropRight(partitioned), partition(last(partitioned), predicate)),
        [elements],
    );
}

export { partitionMulti };
