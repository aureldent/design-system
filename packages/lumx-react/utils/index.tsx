export * from './calculatePopoverPlacement';

export * from './isInViewPort';
