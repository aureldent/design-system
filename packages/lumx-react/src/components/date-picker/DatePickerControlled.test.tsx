import React, { ReactElement } from 'react';

import moment from 'moment';

import { mount, shallow } from 'enzyme';
import 'jest-enzyme';

import { ICommonSetup, Wrapper } from '@lumx/react/testing/utils';

import { DatePickerControlled, DatePickerControlledProps } from './DatePickerControlled';

Date.now = jest.fn(() => new Date(Date.UTC(2017, 1, 14)).valueOf());

/////////////////////////////

/**
 * Define the overriding properties waited by the `setup` function.
 */
type ISetupProps = Partial<DatePickerControlledProps>;

/**
 * Defines what the `setup` function will return.
 */
interface ISetup extends ICommonSetup {
    props: ISetupProps;
}

/////////////////////////////

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 *
 * @param props  The props to use to override the default props of the component.
 * @param     [shallowRendering=true] Indicates if we want to do a shallow or a full rendering.
 * @return      An object with the props, the component wrapper and some shortcut to some element inside of the
 *                       component.
 */
const setup = ({ ...propsOverrides }: ISetupProps = {}, shallowRendering: boolean = true): ISetup => {
    const props: DatePickerControlledProps = {
        locale: 'fr',
        monthOffset: 0,
        onChange: jest.fn(),
        onNextMonthChange: jest.fn(),
        onPrevMonthChange: jest.fn(),
        today: moment(),
        value: moment(),
        ...propsOverrides,
    };

    const renderer: (el: ReactElement) => Wrapper = shallowRendering ? shallow : mount;

    const wrapper: Wrapper = renderer(<DatePickerControlled {...props} />);

    return {
        props,
        wrapper,
    };
};

describe(`<${DatePickerControlled.displayName}>`, (): void => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', (): void => {
        it('should render correctly', (): void => {
            const { wrapper } = setup();
            expect(wrapper).toMatchSnapshot();
        });
    });

    /////////////////////////////

    // 2. Test defaultProps value and important props custom values.
    describe('Props', (): void => {
        // Nothing to do here.
    });

    /////////////////////////////

    // 3. Test events.
    describe('Events', (): void => {
        // Nothing to do here.
    });
    /////////////////////////////

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', (): void => {
        // Nothing to do here.
    });

    /////////////////////////////

    // 5. Test state.
    describe('State', (): void => {
        // Nothing to do here.
    });
});
