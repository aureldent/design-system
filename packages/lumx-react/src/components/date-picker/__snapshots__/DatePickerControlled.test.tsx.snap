// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePickerControlled> Snapshots and structure should render correctly 1`] = `
<div
  className="lumx-date-picker"
>
  <Toolbar
    after={
      <IconButton
        emphasis="low"
        icon="M8.59,16.58L13.17,12L8.59,7.41L10,6L16,12L10,18L8.59,16.58Z"
        onClick={[MockFunction]}
      />
    }
    before={
      <IconButton
        emphasis="low"
        icon="M15.41,16.58L10.83,12L15.41,7.41L14,6L8,12L14,18L15.41,16.58Z"
        onClick={[MockFunction]}
      />
    }
    className="lumx-date-picker__toolbar"
    label={
      <span
        className="lumx-date-picker__month"
      >
        f√©vrier 2017
      </span>
    }
  />
  <div
    className="lumx-date-picker__calendar"
  >
    <div
      className="lumx-date-picker__week-days lumx-date-picker__days-wrapper"
    >
      <div
        className="lumx-date-picker__day-wrapper"
        key="1486944000"
      >
        <span
          className="lumx-date-picker__week-day"
        >
          L
        </span>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1487030400"
      >
        <span
          className="lumx-date-picker__week-day"
        >
          M
        </span>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1487116800"
      >
        <span
          className="lumx-date-picker__week-day"
        >
          M
        </span>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1487203200"
      >
        <span
          className="lumx-date-picker__week-day"
        >
          J
        </span>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1487289600"
      >
        <span
          className="lumx-date-picker__week-day"
        >
          V
        </span>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1487376000"
      >
        <span
          className="lumx-date-picker__week-day"
        >
          S
        </span>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1487462400"
      >
        <span
          className="lumx-date-picker__week-day"
        >
          D
        </span>
      </div>
    </div>
    <div
      className="lumx-date-picker__month-days lumx-date-picker__days-wrapper"
    >
      <div
        className="lumx-date-picker__day-wrapper"
        key="1485730800"
      />
      <div
        className="lumx-date-picker__day-wrapper"
        key="1485817200"
      />
      <div
        className="lumx-date-picker__day-wrapper"
        key="1485903600"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            01
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1485990000"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            02
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1486076400"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            03
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1486162800"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            04
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1486249200"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            05
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1486335600"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            06
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1486422000"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            07
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1486508400"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            08
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1486594800"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            09
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1486681200"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            10
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1486767600"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            11
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1486854000"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            12
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1486940400"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            13
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1487026800"
      >
        <button
          className="lumx-date-picker__month-day lumx-date-picker__month-day--is-today"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            14
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1487113200"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            15
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1487199600"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            16
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1487286000"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            17
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1487372400"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            18
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1487458800"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            19
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1487545200"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            20
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1487631600"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            21
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1487718000"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            22
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1487804400"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            23
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1487890800"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            24
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1487977200"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            25
          </span>
        </button>
      </div>
      <div
        className="lumx-date-picker__day-wrapper"
        key="1488063600"
      >
        <button
          className="lumx-date-picker__month-day"
          disabled={false}
          onClick={[Function]}
        >
          <span>
            26
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
